include/cuco/detail/open_addressing/open_addressing_impl.cuh

```diff
@@ -681,6 +682,91 @@ class open_addressing_impl {
     return output_begin + h_num_out;
   }
 
+  /**
+   * @brief Executes a callback on every filled element in the container.
+   *
+   * @note Passes a copy of the filled element to the callback.
+   *
+   * @tparam CallbackOp Unary callback functor or device lambda
+   *
+   * @param callback_op Function to call on every filled element in the container
+   * @param stream CUDA stream used for this operation
+   */
+  template <typename CallbackOp>
+  void for_each_async(CallbackOp&& callback_op, cuda::stream_ref stream)
+  {
+    std::size_t temp_storage_bytes = 0;
+    using temp_allocator_type =
+      typename std::allocator_traits<allocator_type>::template rebind_alloc<char>;
+    auto temp_allocator = temp_allocator_type{this->allocator()};
+
+    auto const is_filled = open_addressing_ns::detail::slot_is_filled<has_payload, key_type>{
+      this->empty_key_sentinel(), this->erased_key_sentinel()};
+
+    auto const op =
+      [callback_op, is_filled, storage_ = this->storage_ref()] __device__(auto const idx) {
+        auto const window_idx = idx / storage_ref_type::window_size;
+        auto const intra_idx  = idx % storage_ref_type::window_size;
+        auto const slot       = storage_[window_idx][intra_idx];
+
+        if (is_filled(slot)) { callback_op(slot); }
+      };
+
+    CUCO_CUDA_TRY(cub::DeviceFor::ForEachN(nullptr,
+                                           temp_storage_bytes,
+                                           thrust::make_counting_iterator(static_cast<size_t>(0)),
+                                           this->capacity(),
+                                           op,
+                                           stream.get()));
+
+    // Allocate temporary storage
+    auto d_temp_storage = temp_allocator.allocate(temp_storage_bytes);
```

Comment:
This allocation (unless it's stream-ordered/async allocator) prevents this function from being async w.r.t. the host. Does `ForEachN` even need a temp allocation?

Reply:
> Does ForEachN even need a temp allocation?

For my education, if we know in advance that the algorithm won't use temporary data, can we bypass the "allocate" step and just pass `nullptr` to `d_temp_storage`?

Reply:
If we pass `nullptr` wouldn't that trigger the dummy call again, i.e. only computing the temp memory size and not doing any actual work?

Reply:
@srinivasyadav18 It got to my attention that there is [another overload](https://nvidia.github.io/cccl/cub/api/structcub_1_1DeviceFor.html?highlight=devicefor#_CPPv4I000EN3cub9DeviceFor12ForEachCopyNE11cudaError_t21RandomAccessIteratorT9NumItemsT3OpT12cudaStream_t) of `ForEachCopyN` that doesn't require the temp storage parameter. If we switch to that one we should be golden.